### PR TITLE
[GStreamer] Handle latency messages

### PR DESCRIPTION
**Pull request by philn**
#### 57191c0bb3eefcf88bbc54b10e4903420c862cbb
<pre>
[GStreamer] Handle latency messages
<a href="https://bugs.webkit.org/show_bug.cgi?id=248813">https://bugs.webkit.org/show_bug.cgi?id=248813</a>

Reviewed by Xabier Rodriguez-Calvar.

When a latency message is received on the bus the application should recalculate and distribute a
new latency. Also included in this patch is a drive-by improvement in state-change message handling
in the audio file reader, where we process only the notification from the top-level pipeline.

* Source/WebCore/platform/audio/gstreamer/AudioDestinationGStreamer.cpp:
(WebCore::AudioDestinationGStreamer::handleMessage):
* Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp:
(WebCore::AudioFileReader::handleMessage):
* Source/WebCore/platform/audio/gstreamer/AudioSourceProviderGStreamer.cpp:
(WebCore::AudioSourceProviderGStreamer::AudioSourceProviderGStreamer):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::connectSimpleBusMessageCallback):
* Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.cpp:
(WebCore::ImageDecoderGStreamer::InnerDecoder::handleMessage):

Canonical link: <a href="https://commits.webkit.org/257788@main">https://commits.webkit.org/257788@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/b696958f28e270ecea0049bd02ef2d6666004adb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100022 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9193 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33103 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109369 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169610 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104021 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10073 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86547 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92469 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107268 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105790 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90895 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34329 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89518 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/22288 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2985 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23803 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2944 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/46175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9078 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43284 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5346 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4794 "Built successfully") | | | 
| | | | |

### CODE REVIEW

**Comment by philn** [comment]
Ping

File: Source/WebCore/platform/audio/gstreamer/AudioSourceProviderGStreamer.cpp [hunk 1]
```diff
@@ -122,18 +123,23 @@ AudioSourceProviderGStreamer::AudioSourceProviderGStreamer(MediaStreamTrackPriva
     ASSERT(bus);
 
     gst_bus_set_sync_handler(bus.get(), [](GstBus*, GstMessage* message, gpointer userData) -> GstBusSyncReply {
-        auto* decodebin = GST_ELEMENT_CAST(userData);
-        if (GST_MESSAGE_TYPE(message) != GST_MESSAGE_STREAM_COLLECTION || GST_MESSAGE_SRC(message) != GST_OBJECT_CAST(decodebin)) {
+        auto scopeExit = makeScopeExit([&] {
```
Comment (zdobersek) [inline Source/WebCore/platform/audio/gstreamer/AudioSourceProviderGStreamer.cpp]: Adopt `message` into a GRefPtr, would that work?

Comment (philn) [reply Source/WebCore/platform/audio/gstreamer/AudioSourceProviderGStreamer.cpp]: Yeah, I'll update the patch, thanks for the review!

Comment (calvaris) [inline Source/WebCore/platform/audio/gstreamer/AudioSourceProviderGStreamer.cpp]: I think the proper way of doing this would be adopting the message into a GRefPtr.

File: Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp [hunk 2]
```diff
@@ -239,22 +239,22 @@ void AudioFileReader::handleMessage(GstMessage* message)
         gst_element_set_state(m_pipeline.get(), GST_STATE_NULL);
         m_runLoop.stop();
         break;
-    case GST_MESSAGE_STATE_CHANGED:
-        if (GST_MESSAGE_SRC(message) == GST_OBJECT(m_pipeline.get())) {
-            GstState oldState, newState, pending;
-            gst_message_parse_state_changed(message, &oldState, &newState, &pending);
+    case GST_MESSAGE_STATE_CHANGED: {
+        if (GST_MESSAGE_SRC(message) != GST_OBJECT(m_pipeline.get()))
```
Comment (calvaris) [inline Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp:243]: Btw, please mention this drive-by improvement in the changelog

**Review by calvaris — APPROVED** [review]
